File: GLM/Part 3 - Binomial/drawGLM_CI.py
import pandas as pd
import numpy as np
import statsmodels.api as sm
import matplotlib.pyplot as plt
from scipy.stats import norm

# Read the data
DF = pd.read_csv("example_data_ratios.csv")

# Fit the model
n = 10
GLM = sm.GLM(np.column_stack((DF['correct'], n - DF['correct'])),
             sm.add_constant(DF['time']),
             family=sm.families.Binomial(link=sm.families.links.logit())).fit()

# Enter the estimates
intercept = GLM.params[0]
slope = GLM.params[1]

# Create a large sequence of numbers along the x-axis
x = np.linspace(0, 20, 1000)

# The following gives us two things:
# - Predictions of lambda for every value of x
# - A standard error at that value of x
pred = GLM.get_prediction(exog=sm.add_constant(x)).linpred

# Combined, those two can create a 95% CI on the scale of the linear predictor.
# On this scale (not on the original scale!), we can use a normal-approximation.
eta    = pred.predicted_mean
se_fit = pred.se_mean

# Compute the confidence intervals on the scale of the linear predictor
eta_lwr = eta + norm.ppf(0.025) * se_fit
eta_upr = eta + norm.ppf(0.975) * se_fit

# Compute the resulting values on the y-axis (and multiply by n)
y = 1 / (np.exp(-eta) + 1) * n
y_lwr = 1 / (np.exp(-eta_lwr) + 1) * n
y_upr = 1 / (np.exp(-eta_upr) + 1) * n

# Plot the resulting model
plt.plot(x, y)

# Add the confidence band
plt.plot(x, y_lwr, linestyle = 'dashed')
plt.plot(x, y_upr, linestyle = 'dashed')

# Show the plot
plt.show()

File: GLM/Part 3 - Binomial/drawGLM_CI.R
# Read the data (Session --> Set Working Directory --> To source file location)
DF <- read.csv("example_data_ratios.csv")

# Fit the model
n   <- 10
GLM <- glm(cbind(correct, n - correct) ~ time,
           family = binomial(link = "logit"), data = DF)

# Enter the estimates
intercept   <- coef(GLM)[1]
slope       <- coef(GLM)[2]
inverselink <- function(eta){
  1 / (exp(-eta) + 1)
}

# Create a large sequence of numbers along the x-axis
x <- seq(0, 20, length.out = 1000)

# The following gives us two things:
#  - Predictions of lambda for every value of x
#  - A standard error at that value of x
pred <- predict(GLM, newdata = list(time = x), type = "link", se.fit = TRUE)

# Combined, those two can create a 95% CI on the scale of linear predictor.
# On this scale (not on the original scale!), we can use a normal-approximation.
eta     <- pred$fit
eta_lwr <- pred$fit + qnorm(0.025) * pred$se.fit
eta_upr <- pred$fit + qnorm(0.975) * pred$se.fit
  
# Compute the resulting values on the y-axis (and multiply by n)
y     <- inverselink(eta) * n
y_lwr <- inverselink(eta_lwr) * n
y_upr <- inverselink(eta_upr) * n

# Plot the resulting model
plot(x, y, type = "l")

# Add the confidence band
lines(x, y_lwr, lty = "dashed")
lines(x, y_upr, lty = "dashed")
